prov/rxm/src/rxm_conn.c

```diff
@@ -133,62 +128,110 @@ static void rxm_send_queue_close(struct rxm_send_queue *send_queue)
-			  &rxm_conn->posted_rx_list);
-	rxm_conn->saved_msg_ep = rxm_conn->msg_ep;
-	FI_DBG(&rxm_prov, FI_LOG_EP_CTRL,
-	       "Saved MSG EP fid for further deletion in main thread\n");
+
+	rxm_conn->close_data = calloc(1, sizeof(*rxm_conn->close_data));
+	if (!rxm_conn->close_data) {
+		rxm_conn_close_msg_ep(rxm_ep, rxm_conn);
+		return;
+	}
+
+	rxm_conn->close_data->msg_ep = rxm_conn->msg_ep;
+	dlist_init(&rxm_conn->close_data->posted_rx_list);
+	dlist_splice_tail(&rxm_conn->close_data->posted_rx_list, &rxm_conn->posted_rx_list);
 	rxm_conn->msg_ep = NULL;
+	/* Now we ready to create new MSG EP */
 }
 
 static void rxm_conn_free(struct util_cmap_handle *handle)
 {
 	struct rxm_ep *rxm_ep = container_of(handle->cmap->ep, struct rxm_ep, util_ep);
 	struct rxm_conn *rxm_conn = container_of(handle, struct rxm_conn, handle);
 
-	/* This handles case when saved_msg_ep wasn't closed */
-	if (rxm_conn->saved_msg_ep) {
-		rxm_ep_cleanup_posted_rx_list(rxm_ep, &rxm_conn->saved_posted_rx_list);
-		if (fi_close(&rxm_conn->saved_msg_ep->fid))
-			FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
-				"Unable to close saved msg_ep\n");
+	if (handle->cmap->cm_thread_state == UTIL_CM_THREAD_OPEN) {
+		struct rxm_cmd_conn_free_data conn_free_data = {
+			.type = RXM_CMD_CONN_FREE,
+			.rxm_conn = rxm_conn,
+		};
+		struct util_cmap_cmd *cmd =
+			calloc(1, sizeof(struct util_cmap_cmd) +
+				  sizeof(struct rxm_cmd_conn_free_data));
+		if (!cmd)
+			goto conn_free;
+		cmd->type = UTIL_CMAP_CMD_INTERNAL;
+		memcpy(cmd->data, &conn_free_data, sizeof(conn_free_data));
+		dlist_ts_insert_tail(&cmd->entry, &rxm_ep->util_ep.cmap->cmd_queue);
+		rxm_ep->util_ep.cmap->cmd_write++;
+		rxm_conn_wake_up_wait_obj(rxm_ep);
+		return;
 	}
 
-	if (!rxm_conn->msg_ep)
-		return;
-	rxm_ep_cleanup_posted_rx_list(rxm_ep, &rxm_conn->posted_rx_list);
-	/* Assuming fi_close also shuts down the connection gracefully if the
-	 * endpoint is in connected state */
-	if (fi_close(&rxm_conn->msg_ep->fid))
-		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL, "Unable to close msg_ep\n");
-	FI_DBG(&rxm_prov, FI_LOG_EP_CTRL, "Closed msg_ep\n");
-	rxm_conn->msg_ep = NULL;
+conn_free:
+
+	rxm_conn_close_msg_ep(rxm_ep, rxm_conn);
 	rxm_send_queue_close(&rxm_conn->send_queue);
 
-	free(container_of(handle, struct rxm_conn, handle));
+	free(rxm_conn);
 }
 
-static void rxm_conn_connected_handler(struct util_cmap_handle *handle)
+static void
+rxm_conn_process_connected(struct rxm_ep *rxm_ep, struct rxm_conn *rxm_conn)
 {
-	struct rxm_ep *rxm_ep = container_of(handle->cmap->ep, struct rxm_ep, util_ep);
-	struct rxm_conn *rxm_conn = container_of(handle, struct rxm_conn, handle);
+	struct util_cmap_cmd *cmd_connected;
+	ssize_t ret;
+	struct rxm_cmd_connect_data connected_data = {
+		.type = RXM_CMD_CONNECTED,
+		.rxm_conn = rxm_conn,
+	};
 
-	if (!rxm_conn->saved_msg_ep)
+	cmd_connected = calloc(1, sizeof(struct util_cmap_cmd) +
+				  sizeof(struct rxm_cmd_connect_data));
+	if (!cmd_connected) {
```

Comment:
This should never happen. We should assert(0) / return an error instead of working around.

Reply:
Yes, this should never happen. But I think that using fallback mechanism (perform handling of the command here) is good  here.